pkg/comp-functions/functions/common/billing_service.go

```diff
@@ -0,0 +1,196 @@
+package common
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	xfnproto "github.com/crossplane/function-sdk-go/proto/v1"
+	xkube "github.com/vshn/appcat/v4/apis/kubernetes/v1alpha2"
+	vshnv1 "github.com/vshn/appcat/v4/apis/vshn/v1"
+	"github.com/vshn/appcat/v4/pkg/comp-functions/runtime"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	// BillingNamespace is the namespace where BillingService CRs are created
+	BillingNamespace = "syn-appcat"
+	// DefaultKeepAfterDeletion is the default number of days to keep billing records after deletion
+	// it si overwritten by the component value appcat.billing.customResourceDeletionAfter
+	DefaultKeepAfterDeletion = 365
+)
+
+// BillingServiceOptions contains customization options for creating a BillingService CR
+type BillingServiceOptions struct {
+	// ResourceNameSuffix is appended to comp.GetName() to form the resource name (e.g., "-billing-service", "-addon-collabora")
+	ResourceNameSuffix string
+	// ProductID overrides the auto-generated productID based on service type and sla
+	ProductID string
+	// Size overrides the replica count for billing purposes
+	Size string
+	// AdditionalLabels are added to the BillingService CR labels
+	AdditionalLabels map[string]string
+}
+
+// CreateOrUpdateBillingService creates or updates a BillingService CR for the given service instance.
+// The salesOrder is populated from the composition for APPUiO Managed
+// The Organisation is populated from claim namespace for APPUiO Cloud
+// The productID is constructed as: appcat-vshn-{service}-{sla}
+func CreateOrUpdateBillingService(ctx context.Context, svc *runtime.ServiceRuntime, comp InfoGetter) *xfnproto.Result {
+	return CreateOrUpdateBillingServiceWithOptions(ctx, svc, comp, BillingServiceOptions{
+		ResourceNameSuffix: "-billing-service",
+	})
+}
+
+// CreateOrUpdateBillingServiceWithOptions creates or updates a BillingService CR with custom options used for AddOns
+func CreateOrUpdateBillingServiceWithOptions(ctx context.Context, svc *runtime.ServiceRuntime, comp InfoGetter, opts BillingServiceOptions) *xfnproto.Result {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("Creating or updating BillingService", "service", comp.GetName())
+
+	// Skip new billing if disabled
+	if svc.Config.Data["billingEnabled"] == "false" {
+		return runtime.NewNormalResult(fmt.Sprintf("new billing not enabled, skipping... %s", comp.GetName()))
```

Comment:
```suggestion
		return runtime.NewNormalResult(fmt.Sprintf("Bbilling not enabled, skipping... %s", comp.GetName()))
```
Makes no sense to call it "new" as it won't be "new" for long :-)

Reply:
yes true, they will not run in parallel, but the code will be updated here in the next ticket